splay/src/main/scala/org/scalajs/benchmark/splay/Splay.scala

```diff
@@ -0,0 +1,439 @@
+   * key, the last node on the search path is moved to the top of the
+   * tree. This is the simplified top-down splaying algorithm from:
+   * "Self-adjusting Binary Search Trees" by Sleator and Tarjan
+   *
+   * @param {number} key Key to splay the tree on.
+   * @private
+   */
+  private def splay(key: Int): Unit = {
+    if (isEmpty) {
+      // Do nothing
+    } else {
+      // Create a dummy node.  The use of the dummy node is a bit
+      // counter-intuitive: The right child of the dummy node will hold
+      // the L tree of the algorithm.  The left child of the dummy node
+      // will hold the R tree of the algorithm.  Using a dummy node, left
+      // and right will always be nodes and we avoid special cases.
+      val dummy = Node.newDummy()
+      var left = dummy
+      var right = dummy
+      var current = root
+      var break = false
+      while (!break) {
+        if (key < current.key) {
+          if (current.left == null) {
+            break = true
+          } else {
+            if (key < current.left.key) {
+              // Rotate right.
+              val tmp = current.left
+              current.left = tmp.right
+              tmp.right = current
+              current = tmp
+              if (current.left == null) {
+                break = true
+              }
+            }
+            if (!break) {
+              // Link right.
+              right.left = current
+              right = current
+              current = current.left
+            }
+          }
+        } else if (key > current.key) {
+          if (current.right == null) {
+            break = true
+          } else {
+            if (key > current.right.key) {
+              // Rotate left.
+              val tmp = current.right
+              current.right = tmp.left
+              tmp.left = current
+              current = tmp
+              if (current.right == null) {
+                break = true
+              }
+            }
+            if (!break) {
+              // Link left.
+              left.right = current
+              left = current
+              current = current.right
+            }
+          }
+        } else {
+          break = true
+        }
+      }
+      // Assemble.
+      left.right = current.left
+      right.left = current.right
+      current.left = dummy.right
+      current.right = dummy.left
+      root = current
+    }
+  }
+
+  /**
+   * Constructs a Splay tree node.
+   *
+   * @param {number} key Key.
+   * @param {*} value Value.
+   */
+  class Node(val key: Int, val value: T) {
+
+    /**
+     * @type {SplayTree.Node}
+     */
+    var left: Node = null
+
+    /**
+     * @type {SplayTree.Node}
+     */
+    var right: Node = null
+
+    /**
+     * Performs an ordered traversal of the subtree starting at
+     * this SplayTree.Node.
+     *
+     * @param {function(SplayTree.Node)} f Visitor function.
+     * @private
+     */
+    def traverse(f: Node => Unit): Unit = {
```

Comment:
One thing to try: type `f` as `js.Function1[Node, Unit]` to test whether that is a performance bottleneck.